pkg/diff/diff.go

```diff
@@ -487,140 +528,187 @@ func (t *TableDiff) compareChecksum(ctx context.Context, chunk *ChunkRange) (boo
 }
 
 func (t *TableDiff) compareRows(ctx context.Context, chunk *ChunkRange) (bool, error) {
-	sourceRows := make(map[string][]map[string]*dbutil.ColumnData)
+	beginTime := time.Now()
+
+	sourceRows := make(map[int]*sql.Rows)
+	sourceMap := make(map[int]bool)
 	args := utils.StringsToInterfaces(chunk.Args)
 
 	targetRows, orderKeyCols, err := getChunkRows(ctx, t.TargetTable.Conn, t.TargetTable.Schema, t.TargetTable.Table, t.TargetTable.info, chunk.Where, args, t.Collation)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
-
-	// judge rows have all order keys to avoid panic
-	if len(targetRows) > 0 {
-		if !rowContainsCols(targetRows[0], orderKeyCols) {
-			return false, errors.Errorf("%s.%s.%s's data don't contain all keys %v", t.TargetTable.InstanceID, t.TargetTable.Schema, t.TargetTable.Table, orderKeyCols)
-		}
-	}
+	defer targetRows.Close()
 
 	for i, sourceTable := range t.SourceTables {
 		rows, _, err := getChunkRows(ctx, sourceTable.Conn, sourceTable.Schema, sourceTable.Table, sourceTable.info, chunk.Where, args, t.Collation)
 		if err != nil {
 			return false, errors.Trace(err)
 		}
+		defer rows.Close()
 
-		// judge rows have all order keys to avoid panic
-		if len(rows) > 0 {
-			if !rowContainsCols(rows[0], orderKeyCols) {
-				return false, errors.Errorf("%s.%s.%s's data don't contain all keys %v", sourceTable.InstanceID, sourceTable.Schema, sourceTable.Table, orderKeyCols)
-			}
-		}
-
-		sourceRows[fmt.Sprintf("source-%d", i)] = rows
+		sourceRows[i] = rows
+		sourceMap[i] = false
```

Comment:
Can you give it a meaningful name now?

Reply:
rename to `sourceHaveData` in https://github.com/pingcap/tidb-tools/pull/370/commits/33cdde062dfcf499e2cc4e2a48abcf4030123832